form/data_mappers.rst

```diff
@@ -0,0 +1,197 @@
+    using a data mapper, you should configure the ``empty_data`` with a closure
+    as described in
+    :ref:`How to Configure empty Data for a Form Class <forms-empty-data-closure>`.
+
+The red, green and blue form fields have to be mapped to the constructor
+arguments and the ``Color`` instance has to be mapped to red, green and blue
+form fields. Recognize a familair pattern? It's time for a data mapper!
+
+.. code-block:: php
+
+    // src/AppBundle/Form/DataMapper/ColorMapper.php
+    namespace AppBundle\Form\DataMapper;
+
+    use AppBundle\Colors\Color;
+    use Symfony\Component\Form\DataMapperInterface;
+    use Symfony\Component\Form\Exception\UnexpectedTypeException;
+
+    class ColorMapper extends DataMapperInterface
+    {
+        public function mapDataToForms($data, $forms)
+        {
+            // there is no data yet, a new color will be created
+            if (null === $data) {
+                return;
+            }
+
+            // invalid data type, this message will not be shown to the user (see below)
+            if (!$data instanceof Color) {
+                throw new UnexpectedTypeException($data, Color::class);
+            }
+
+            $forms = iterator_to_array($forms);
+
+            // set form field values
+            $forms['red']->setData($data->getRed());
+            $forms['green']->setData($data->getGreen());
+            $forms['blue']->setData($data->getBlue());
+        }
+
+        public function mapFormsToData($forms, &$data)
+        {
+            $forms = iterator_to_array($forms);
+
+            // get form field values
+            $red = $forms['red']->getData();
+            $green = $forms['green']->getData();
+            $blue = $forms['blue']->getData();
+
+            $data = new Color($red, $green, $blue);
+        }
+    }
+
+All data mappers have to implement :class:`Symfony\\Component\\Form\\DataMapperInterface`.
+This interface contains two methods:
+
+``mapDataToForms()``
+    Recieves the current view data and a :phpclass:`RecursiveIteratorIterator`
+    with all form fields. It's task is to map the view data to the form fields.
+    The view data might be ``null`` if there is no data yet (e.g. in a "Create
+    new color" form);
+
+``mapFormsToData($forms, &$data)``
+    Maps the form field values to the ``$data`` property, which is changed by
+    reference. It recieves the form fields in a :phpclass:`RecursiveIteratorIterator`
+    and the view data.
+
+When an errors occurs, a
+:class:`Symfony\\Component\\Form\\Exception\\TransformationFailedException` or
+one of its subclasses should be thrown. The exception message won't be shown to
+the user. Instead, the ``invalid_message`` option will be used.
+
+.. caution::
+
+    The data passed to the mapper is *not yet validated*. This means that your
+    objects should allow being created in an invalid state in order to produce
+    user-friendly errors in the form.
```

Comment:
This is also why scalar type hints in PHP7 cause some troubles. The form maps submitted data before validating its underlying data.

What we have to do in such case, is to validate data before mapping it ourselves :( I think he talked about it in its last conferences (suggestion to use custom data mappers for PHP 7).

Not sure what's the best way to approach this problem though, a "caution"  note might be enough after all.